src/Components/Web.JS/src/Boot.Web.ts

```diff
@@ -9,32 +9,44 @@
 //    of interactive components
 
 import { DotNet } from '@microsoft/dotnet-js-interop';
-import { startCircuit } from './Boot.Server.Common';
-import { startWebAssembly } from './Boot.WebAssembly.Common';
+import { setCircuitOptions, startCircuit } from './Boot.Server.Common';
+import { loadWebAssemblyPlatform, setWebAssemblyOptions, startWebAssembly } from './Boot.WebAssembly.Common';
 import { shouldAutoStart } from './BootCommon';
 import { Blazor } from './GlobalExports';
 import { WebStartOptions } from './Platform/WebStartOptions';
 import { attachStreamingRenderingListener } from './Rendering/StreamingRendering';
 import { NavigationEnhancementCallbacks, attachProgressivelyEnhancedNavigationListener, isPerformingEnhancedPageLoad } from './Services/NavigationEnhancement';
-import { WebAssemblyComponentDescriptor } from './Services/ComponentDescriptorDiscovery';
-import { ServerComponentDescriptor } from './Services/ComponentDescriptorDiscovery';
-import { RootComponentManager } from './Services/RootComponentManager';
-import { WebRendererId } from './Rendering/WebRendererId';
+import { ComponentDescriptor } from './Services/ComponentDescriptorDiscovery';
+import { RootComponentManager, attachAutoModeResolver } from './Services/RootComponentManager';
 import { DescriptorHandler, attachComponentDescriptorHandler, registerAllComponentDescriptors } from './Rendering/DomMerging/DomSync';
 import { waitForRendererAttached } from './Rendering/WebRendererInteropMethods';
+import { WebRendererId } from './Rendering/WebRendererId';
+
+enum WebAssemblyLoadingState {
+  None = 0,
+  Loading = 1,
+  Loaded = 2,
+  Starting = 3,
+  Started = 4,
+}
 
 let started = false;
-let webStartOptions: Partial<WebStartOptions> | undefined;
+let hasCircuitStarted = false;
+let webAssemblyLoadingState = WebAssemblyLoadingState.None;
+let autoModeWebAssemblyTimeoutMilliseconds: number;
+let autoModeTimeoutState: undefined | 'waiting' | 'timed out';
 
-const circuitRootComponents = new RootComponentManager(WebRendererId.Server);
-const webAssemblyRootComponents = new RootComponentManager(WebRendererId.WebAssembly);
+const rootComponentManager = new RootComponentManager();
 
 function boot(options?: Partial<WebStartOptions>) : Promise<void> {
   if (started) {
     throw new Error('Blazor has already started.');
   }
   started = true;
-  webStartOptions = options;
+  autoModeWebAssemblyTimeoutMilliseconds = options?.ssr?.autoModeWebAssemblyTimeoutMilliseconds ?? 100;
```

Comment:
> My guess is the reason you did it this way is that when we moved...

Yes, that's exactly right. This was meant as a workaround because information about which files are already cached is hidden from Blazor. I'm totally fine with filing an issue to make any necessary `dotnet.js` API changes to make this possible.